hail/python/hail/methods/statgen.py

```diff
@@ -1480,13 +1480,23 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
         field = Env.get_uid()
         mt = mt.select_entries(**{field: entry_expr})
     mt = mt.select_cols().select_rows().select_globals()
-
-    r = Env.hail().methods.PCA.apply(mt._jmt, field, k, compute_loadings)
-    scores = Table._from_java(Env.hail().methods.PCA.scoresTable(mt._jmt, r._2()))
-    loadings = from_option(r._3())
-    if loadings:
-        loadings = Table._from_java(loadings)
-    return jiterable_to_list(r._1()), scores, loadings
+    
+    t = Table(MatrixToTableApply(mt._mir, {
+        'name': 'PCA',
+        'entryField': field,
+        'k': k,
+        'computeLoadings': compute_loadings
+    }))
+
+    g = hl.eval(t.index_globals())
+
+    col_keys = mt.col_key.collect()
+    scores = [k.annotate(scores=s) for k, s in zip(col_keys, g.scores)]
+    scores = hl.Table.parallelize(scores, mt.col_key.dtype._insert_field('scores', dtype('array<float64>')), key=list(mt.col_key))
```

Comment:
actually, I think that the relational function should put the col keys into globals so that all you need to do here is `TableParallelize(GetField(TableGetGLobals(..., 'scores'))`

Reply:
Not sure I understand.  Is the problem here we're run PCA twice, once for the scores/eigenvalues and again if you compute the loadings and access them?  If so, I'm not sure how making the collect lazy fixes this.  It just defers the evaluation of the globals, but you've still got two pipelines.

We need to cache the output of PCA, but we also need to make cache cache the globals for a table once they get computed.  Right now, cache only applies to the RDD part.  Hmm.

Reply:
the problem is you're evaluating `mt` twice, which should be easy to fix by including the col keys in the resulting table.

Reply:
persist() will cache the globals, since it will produce a TableLiteral right now.

Reply:
Oh, you're right, persist does cache the globals.  OK, got it, will make this change.

Reply:
OK, I made the changes (but didn't test locally, so there might be some tweaks to get it all working again).